nautobot/core/apps/__init__.py

```diff
@@ -0,0 +1,247 @@
+from django.apps import AppConfig
+from django.urls import reverse
+from operator import getitem
+from collections import OrderedDict
+
+from nautobot.extras.plugins.utils import import_object
+from nautobot.extras.registry import registry
+from nautobot.utilities.choices import ButtonActionColorChoices
+
+
+registry["nav_menu"] = {"tabs": {}}
+
+
+class NautobotConfig(AppConfig):
+    """
+    Custom AppConfig for Nautobot application.
+
+    Adds functionality to generate the HTML navigation menu using `navigation.py` files from nautbot
+    applications.
+    """
+
+    menu_tabs = "navigation.menu_tabs"
+
+    def ready(self):
+        """
+        Ready function initiates the import application.
+
+        As the application can be loaded from either the root of the app or `apps.py`, the __module__
+        name needs to be modified to all the ready function to import the menu items.
+
+        To import the menu items the `.apps` needs to be removed from the __module__ name if it
+        exists.
+
+        Examples:
+            Loaded from application root:   `nautobot.circuits`
+            Loaded from `app.py`:           `nautobot.circuits.apps`
+        """
+        if self.__module__.endswith(".apps"):
+            module_location = self.__module__[:-5]
+        else:
+            module_location = self.__module__
+        menu_items = import_object(f"{module_location}.{self.menu_tabs}")
+        if menu_items is not None:
+            register_menu_items(menu_items)
+
+
+def register_menu_items(tab_list):
+    """
+    Using the imported object a dictionary is either created or updated with objects to create
+    the navbar.
+
+    The dictionary is built from four key objects, NavMenuTab, NavMenuGroup, NavMenuItem and
+    NavMenuButton. The Django template then uses this dictionary to generate the navbar HTML.
+    """
+    for nav_tab in tab_list:
+        if nav_tab.name not in registry["nav_menu"]["tabs"]:
+            registry["nav_menu"]["tabs"][nav_tab.name] = {
+                "weight": nav_tab.weight,
+                "groups": {},
+                "permissions": [],
+            }
+
+        tab_perms = []
+        registry_groups = registry["nav_menu"]["tabs"][nav_tab.name]["groups"]
+        for group in nav_tab.groups:
+            if group.name not in registry_groups:
+                registry_groups[group.name] = {
+                    "weight": group.weight,
+                    "items": {},
+                    "permissions": [],
+                }
+
+            group_perms = []
+            for item in group.items:
+                if item.link not in registry_groups[group.name]["items"]:
+                    registry_groups[group.name]["items"][item.link] = {
+                        "link_text": item.link_text,
+                        "weight": item.weight,
+                        "buttons": item.buttons,
```

Comment:
Seems like we should have another nested for loop (`for button in item.buttons:`) to handle buttons so as to be able to update buttons on an existing item and allow button weights to be taken into account.